### PR TITLE
[WIP] [ugly commits] add config observation

### PR DESCRIPTION
**Pull request by stlaz**
currently probably quite fragile, many FIXMEs and TODOs

The previous resource sync for the resources from the oauth config was replaces by syncing it from the configobserver and it seems to work, just as well as does the slightly reworked logic about how mounts are placed in the pods

- [ ] add observers for the fields of the authentication object
- [ ] add observation for whatever we need from the consoles
- [ ] add observation for whatever we need from the infrastructures object
- [ ] make the deployment use the observed config
- [ ] ?
- [ ] profit

### CODE REVIEW

**Comment by stlaz** [comment]
did some very wild rebasing to make this reviewable. Given that github does not respect logical ordering of commits but does it chronologically instead, any reviewer might want to pull the changes and review locally.

File: pkg/operator2/configobservation/oauth/observe_idps_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+)
+
+type mockResourceSyncer struct {
+	t      *testing.T
+	synced map[string]string
+}
+
+func (rs *mockResourceSyncer) SyncConfigMap(destination, source resourcesynccontroller.ResourceLocation) error {
```
Comment (marun) [inline pkg/operator2/configobservation/oauth/observe_idps_test.go]: (No action required) Would it be make sense to avoid mocking by separating api calls from logic, e.g.

```golang

type idpResult {
  currentSyncData  *datasync.ConfigSyncData
  observedSyncData *datasync.ConfigSyncData
  observedConfig   map[string]interface{}
  eventMessage     string
}

// With dependencies on the API, this function is ideally validated in e2e
func ObserveIdentityProviders(genericlisters configobserver.Listers, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, []error) {
    listers := genericlisters.(configobservation.Listers)
    resourceSyncer := genericlisters.ResourceSyncer()

    oauthConfig, _ := listers.OAuthLister.Get("cluster")
    ... 

    result, _ := computeIdpResult(oathConfig, previouslyObservedConfig)
    ...

    datasync.HandleIdPConfigSync(resourceSyncer, result.currentSyncData, result.observedSyncData)
    
    if len(result.eventMessage) {
        recorder.Eventf("ObserveIdentityProviders", result.eventMessage)
    }

    return result.observedConfig, nil
}

// With no dependencies on the API, the logic of this function is ideally validated with unit testing
func computeIdpResult(oathConfig *v1.OAuth, previousConfig map[string]interface{}) (*idpResult, error) {
  ...
}
```

File: pkg/operator2/configobservation/oauth/observe_tokenconfig.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	currentTokenConfig, _, err := unstructured.NestedMap(previouslyObservedConfig, tokenConfigPath...)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	currentAccessTokenMaxAgeSecondsFloat, _, err := unstructured.NestedFloat64(currentTokenConfig, "accessTokenMaxAgeSeconds")
+	if err != nil {
+		errs = append(errs, err)
+	}
+	currentAccessTokenMaxAgeSeconds := float64(currentAccessTokenMaxAgeSecondsFloat)
```
Comment (marun) [inline pkg/operator2/configobservation/oauth/observe_tokenconfig.go]: Is this conversion necessary given that `NestedFloat64` returns a float64?

File: pkg/operator2/configobservation/oauth/observe_tokenconfig.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/configobservation"
+)
+
+const defaultAccessTokenMaxAgeSeconds = float64(86400) // a day
```
Comment (marun) [inline pkg/operator2/configobservation/oauth/observe_tokenconfig.go]: Given that the golant struct fields are int32 and the value in seconds is fixed point, is it desirable to use a floating point type?

File: pkg/operator2/configobservation/oauth/observe_tokenconfig.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+	observedAccessTokenMaxAgeSeconds := float64(oauthConfig.Spec.TokenConfig.AccessTokenMaxAgeSeconds)
+	if observedAccessTokenMaxAgeSeconds == 0 {
+		observedAccessTokenMaxAgeSeconds = defaultAccessTokenMaxAgeSeconds
+	}
+
+	observedTokenConfigFieldMap := map[string]interface{}{}
+	unstructured.SetNestedField(observedTokenConfigFieldMap, observedAccessTokenMaxAgeSeconds, "accessTokenMaxAgeSeconds")
+
+	observedAccessTokenInactivityTimeoutSeconds := float64(oauthConfig.Spec.TokenConfig.AccessTokenInactivityTimeoutSeconds)
```
Comment (marun) [inline pkg/operator2/configobservation/oauth/observe_tokenconfig.go]: It appears that this switch statement is just ensuring a value less than 0 is reset to zero, so maybe simplify as follows?

```golang
if val < 0 {
  val = 0 
}
fieldMap["val"] = val
```

File: pkg/operator2/configobservation/oauth/template_conversions.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		return nil, nil, err
+	}
+
+	switch brand {
+	case okdBrand:
+		// do nothing, OAuth server has right branding for this
+
+	case ocpBrand, "dedicated", "online", "azure":
+		// all of these are equivalent to ocp
+		templates = osinv1.OAuthTemplates{
```
Comment (marun) [inline pkg/operator2/configobservation/oauth/template_conversions.go]: Maybe define this outside of the switch statement and use it here and for default?

File: pkg/operator2/configobservation/routersecret/observe_router_secret.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/configobservation"
+)
+
+func ObserveRouterSecret(genericlisters configobserver.Listers, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (map[string]interface{}, []error) {
+	namedCertificatesPath := []string{"servingInfo", "namedCertificates"}
+	listers := genericlisters.(configobservation.Listers)
+	errs := []error{}
+
+	currentNamedCertificates, _, err := unstructured.NestedFieldCopy(previouslyObservedConfig, namedCertificatesPath...)
```
Comment (marun) [inline pkg/operator2/configobservation/routersecret/observe_router_secret.go]: Maybe use NestedSlice?

File: pkg/operator2/configobservation/apiserver/observe_cors.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+var clusterDefaultCORSALlowedOrigins = []string{
+	`//127\.0\.0\.1(:|$)`,
+	`//localhost(:|$)`,
+}
+
+func ObserveAdditionalCORSAllowedOrigins(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	const corsAllowedOriginsPath = "corsAllowedOrigins"
+
+	listers := genericListers.(configobservation.Listers)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/configobservation/apiserver/observe_cors.go]: The cast should be safe (those listers are constant), if it fails, the programmer did not do his job well (happened to me several times while developing this).

File: pkg/operator2/configobservation/apiserver/observe_cors.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	currentCORSSet.Insert(clusterDefaultCORSALlowedOrigins...)
+
+	observedConfig := map[string]interface{}{}
+	apiServer, err := listers.APIServerLister().Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warningf("apiserver.config.openshift.io/cluster: not found")
+		return defaultConfig, errs
+	}
+	if err != nil {
+		return existingConfig, errs
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/configobservation/apiserver/observe_cors.go]: The error should probably be appended, correct.

`existingConfig` is returned because returning `defaultConfig` might rewrite whatever we had even if `err` is just a transient network error, and we would have another rewrite once that is over. I'll add a comment.

File: pkg/operator2/configobservation/apiserver/observe_cors.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	`//127\.0\.0\.1(:|$)`,
+	`//localhost(:|$)`,
+}
+
+func ObserveAdditionalCORSAllowedOrigins(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	const corsAllowedOriginsPath = "corsAllowedOrigins"
+
+	listers := genericListers.(configobservation.Listers)
+	errs := []error{}
+	defaultConfig := map[string]interface{}{}
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/configobservation/apiserver/observe_cors.go]: `SetNestedStringSlice()` does a small little bit more than just that, I'd rather stick with the current code.

File: pkg/operator2/configobservation/oauth/idp_conversions.go [hunk 10]
```diff
@@ -43,7 +46,61 @@ type idpData struct {
 	login     bool
 }
 
-func (c *authOperator) convertProviderConfigToIDPData(providerConfig *configv1.IdentityProviderConfig, syncData *configSyncData, i int) (*idpData, error) {
+// FIXME: will have to return sync data as well, also maybe warnings?
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/configobservation/oauth/idp_conversions.go]: historical artifacts from WIP, will be removed in next revision, thx

File: pkg/operator2/deployment.go [hunk 11]
```diff
@@ -35,61 +41,84 @@ func defaultDeployment(
...
-		{
-			name:      "v4-0-config-system-session",
-			configmap: false,
-			path:      "/var/config/system/secrets/v4-0-config-system-session",
-			keys:      []string{"v4-0-config-system-session"},
+	// FIXME: ignore errors for now, this should probably be done earlier when working with
+	//        the authentication.operator resource anyway
+	idpSyncData, _ := getSyncDataFromOperatorConfig(&operatorConfig.Spec.ObservedConfig)
+
+	for _, data := range append(
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/deployment.go]: legacy choice, this code is now gone thanks to @sanchezl :champagne: \o/

File: pkg/operator2/configobservation/oauth/observe_idps_test.go (hunk 1 above)
Comment (stlaz) [reply pkg/operator2/configobservation/oauth/observe_idps_test.go]: Both seem to bear more or less similar results I guess, I find the mocking to be quite elegant way to test this, but I can see how your approach separates the logical unit outside so that it is easier to raise expectations over it.

File: pkg/operator2/configobservation/oauth/observe_tokenconfig.go (hunk 2 above)
Comment (stlaz) [reply pkg/operator2/configobservation/oauth/observe_tokenconfig.go]: I need to re-check this module, I think I left it at a WIP state...

File: pkg/operator2/configobservation/oauth/observe_tokenconfig.go (hunk 3 above)
Comment (stlaz) [reply pkg/operator2/configobservation/oauth/observe_tokenconfig.go]: glad you asked, this was such a PITA to get right.

The config is really a json that gets deserialized into golang nested interfaces.

When we make this int32 and store as such, the information about it being int is obviously stored.

Eventually, the `configObserver`, which lives in library-go, compares the config it got originally with the one that it got as observed config.

This is where all the fun happens. because the json parser actually likes to think of numbers as of floats prior to them being integers. That means that if we make this integer, same jsons will actually be considered different in their go interpretation even if our comparison here tells you the numbers are equal. That results in hotloop trying to update the `observedConfig` field on same values.

I tried many things, played with the json parser flags, use `json.Number` as the default representation for numbers found in the jsons we have here but nothing except this really worked.

I did a terrible job with code comments about this, I see.

File: pkg/operator2/configobservation/apiserver/observe_cors.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/configobservation/apiserver/observe_cors.go]: on a second thought, I might try to do that

File: pkg/operator2/datasync/syncdata.go [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+
+	switch s.Resource {
+	case ConfigMapType:
+		vol.ConfigMap = &corev1.ConfigMapVolumeSource{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: volName,
+			},
+			Items: items,
+		}
+	default:
```
Comment (marun) [inline pkg/operator2/datasync/syncdata.go:197]: (No action required) Maybe specify SecretType as an explicit case and panic in default?

File: pkg/operator2/datasync/syncdata.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	// idps that are synced have this prefix
+	return fmt.Sprintf("v4-0-config-user-idp-%d-%s", i, field)
+}
+
+func getIDPPath(i int, resource, dest string) string {
+	// root path for IDP data
+	return fmt.Sprintf("/var/config/user/idp/%d/%s/%s", i, resource, dest)
+}
+
+func syncOrDie(syncFunc func(dest, src resourcesynccontroller.ResourceLocation) error, dest, src string) {
```
Comment (marun) [inline pkg/operator2/datasync/syncdata.go:221]: nit: Maybe factor out the common functionality between this and `syncSecretOrDie`?

File: pkg/operator2/datasync/syncdata.go [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+
+type ResourceType string
+
+const (
+	ConfigMapType ResourceType = "configMap"
+	SecretType    ResourceType = "secret"
+)
+
+type sourceData struct {
+	Src      string       `json:"src"`  // name of the source in openshift-config namespace
```
Comment (marun) [inline pkg/operator2/datasync/syncdata.go:31]: nit: why isn't the name of this field just `Name` ? As per further comments on fields in this struct, a good name can largely obviate the need for explanatory comments.

File: pkg/operator2/datasync/syncdata.go [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+	Path     string       `json:"path"` // the mount path that this source is mapped to
+	Key      string       `json:"key"`
+	Resource ResourceType `json:"resource"`
```
Comment (marun) [inline pkg/operator2/datasync/syncdata.go:34]: nit: why isn't the name of this field `ResourceType` or `Type` ?

File: pkg/operator2/datasync/syncdata.go [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+type ResourceType string
+
+const (
+	ConfigMapType ResourceType = "configMap"
+	SecretType    ResourceType = "secret"
+)
+
+type sourceData struct {
+	Src      string       `json:"src"`  // name of the source in openshift-config namespace
+	Path     string       `json:"path"` // the mount path that this source is mapped to
```
Comment (marun) [inline pkg/operator2/datasync/syncdata.go:32]: nit: why isn't the name of this field `MountPath` ?

File: pkg/operator2/datasync/validation.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return []error{err}
+	}
+
+	data, exists := s.Data[src.Key]
+	if !exists {
+		return []error{fmt.Errorf("missing required key: %q", src.Key)}
+	}
+
+	return validators[src.Key](data)
```
Comment (marun) [inline pkg/operator2/datasync/validation.go:39]: Would it make sense to check that the key was found and return an error if not?

File: pkg/operator2/datasync/validation.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return []error{err}
+	}
+
+	data, exists := cm.Data[src.Key]
+	if !exists {
+		return []error{fmt.Errorf("missing required key: %q", src.Key)}
+	}
+
+	return validators[src.Key]([]byte(data))
```
Comment (marun) [inline pkg/operator2/datasync/validation.go:53]: ditto

File: pkg/operator2/datasync/validation.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+	return errs
+}
+
+func validatePrivateKey(pemKey []byte) []error {
+	if len(pemKey) == 0 {
+		return []error{fmt.Errorf("required private key is empty")}
+	}
+	if _, err := keyutil.ParsePrivateKeyPEM(pemKey); err != nil {
+		return []error{fmt.Errorf("failed to parse the private key")}
```
Comment (marun) [inline pkg/operator2/datasync/validation.go:86]: Would it make sense to wrap the original error to aid in troubleshooting?

File: pkg/operator2/datasync/validation.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	errs := []error{}
+	now := time.Now()
+
+	cert, rest := pem.Decode(pemCerts)
+	for ; cert != nil; cert, rest = pem.Decode(rest) {
+		parsed, err := x509.ParseCertificate(cert.Bytes)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to parse a certificate in the chain"))
```
Comment (marun) [inline pkg/operator2/datasync/validation.go:118]: Would it make sense to include the original error?

File: pkg/operator2/datasync/validation_test.go [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func Test_validateSecret(t *testing.T) {
```
Comment (marun) [inline pkg/operator2/datasync/validation_test.go:15]: (No action required) Consider validating the field-agnostic wrapper separately from the actual validation functions to simplify writing the tests (i.e. no indexer/lister required) and make life easier for future maintainers when they attempt to determine coverage.

File: pkg/operator2/datasync/validation_test.go [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+		name    string
+		src     sourceData
+		secrets []*corev1.Secret
+		want    []error
+	}{
+		{
+			name: "secret missing",
+			src:  sourceData{Src: "somesecret", Key: corev1.TLSPrivateKeyKey},
+			want: []error{
+				fmt.Errorf("secret \"somesecret\" not found"),
```
Comment (marun) [inline pkg/operator2/datasync/validation_test.go:26]: (No action required) Consider avoiding a dependency on a specific error message as this style of testing is brittle and will have to be maintained in lock-step with the code.

File: pkg/operator2/datasync/validation.go (hunk 19 above)
Comment (stlaz) [reply pkg/operator2/datasync/validation.go:86]: When it comes to private keys, I'd rather avoid printing errors that I don't directly control.

File: pkg/controllers/configobservation/oauth/idp_conversions.go [hunk 23]
```diff
@@ -50,10 +53,69 @@ type idpData struct {
...
+) ([]interface{}, *datasync.ConfigSyncData, []error) {
+
+	converted := []osinv1.IdentityProvider{}
+	syncData := datasync.NewConfigSyncData()
+	errs := []error{}
+
+	for i, idp := range defaultIDPMappingMethods(identityProviders) {
+		data, err := convertProviderConfigToIDPData(cmLister, secretsLister, &idp.IdentityProviderConfig, syncData, i)
+		if err != nil {
+			klog.Errorf("failed to honor IDP %#v: %v", idp, err)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/idp_conversions.go]: remnants of the ages past, removed

File: pkg/controllers/configobservation/oauth/idp_conversions.go [hunk 24]
```diff
@@ -50,10 +53,69 @@ type idpData struct {
...
+				UseAsLogin:      data.login,
+				MappingMethod:   string(idp.MappingMethod),
+				Provider: runtime.RawExtension{
+					Raw: encodeOrDie(data.provider),
+				},
+			},
+		)
+	}
+
+	// TODO: the following is a bit perverted, I wonder if it could be done differently
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/idp_conversions.go]: I think it's actually fine, I replaced the comment with something more useful.

File: pkg/controllers/configobservation/oauth/observe_idps.go [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+	existingSyncData, err := GetIDPConfigSyncData(existingConfig)
+	if err != nil {
+		return existingConfig, append(errs, err)
+	}
+
+	oauthConfig, err := listers.OAuthLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		// revert to default state, meaning no IdPs
+		klog.Warning("oauth.config.openshift.io/cluster: not found")
+		return map[string]interface{}{}, nil
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_idps.go]: added the errs there

File: pkg/controllers/configobservation/oauth/observe_idps.go [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/configobservation"
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/datasync"
+)
+
+var identityProvidersMounts = []string{"volumesToMount", "identityProviders"}
+
+func ObserveIdentityProviders(genericlisters configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, errs []error) {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_idps.go:19]: The pruning should always happen, and I think it logically belongs within the observation function. We wouldn't have to deal with it at all in theory, but it's a safety precaution based on the observations of the past.

File: pkg/controllers/configobservation/oauth/observe_templates.go [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+
+	var observedTemplates interface{}
+	if templates != nil {
+		convertedBytes, err := json.Marshal(templates)
+		if err != nil {
+			return existingConfig, append(errs, err)
+		}
+
+		if err := json.NewDecoder(bytes.NewBuffer(convertedBytes)).Decode(&observedTemplates); err != nil {
+			return nil, append(errs, fmt.Errorf("decode of observed config failed with error: %v", err))
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_templates.go]: added the existing config instead

File: pkg/operator2/oauth.go [hunk 28]
```diff
@@ -7,21 +7,31 @@ import (
...
 	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
-	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/configobservation"
+)
+
+var (
+	scheme  = runtime.NewScheme()
+	codecs  = serializer.NewCodecFactory(scheme)
+	encoder = codecs.LegacyCodec(osinv1.GroupVersion) // TODO I think there is a better way to do this
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/oauth.go:28]: It's the question old as the repo, but you're right, @sttts might be able to know the answer, I'll ask

File: pkg/controllers/configobservation/oauth/observe_tokenconfig.go [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+	oauthConfig, err := listers.OAuthLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warning("oauth.config.openshift.io/cluster: not found")
+		return map[string]interface{}{
+			"oauthConfig": map[string]interface{}{
+				"tokenConfig": map[string]interface{}{
+					"accessTokenMaxAgeSeconds": defaultAccessTokenMaxAgeSeconds,
+				},
+			},
+		}, nil
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_tokenconfig.go]: added the errors

File: pkg/controllers/configobservation/oauth/template_conversions.go [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+		templates = ocpDefaults
+
+	default:
+		if defaultBrand == ocpBrand {
+			// build-time ocp selection
+			templates = ocpDefaults
+		}
+	}
+
+	// user-configured overrides everything else, individually.
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/template_conversions.go]: The comment is now improved :)

File: pkg/operator2/datasync/syncdata.go [hunk 31]
```diff
@@ -0,0 +1,238 @@
...
+			syncFunc = resourceSyncer.SyncConfigMap
+			newConfigMapNames.Insert(dest)
+		} else {
+			newSecretNames.Insert(dest)
+		}
+
+		syncOrDie(syncFunc, dest, newData.data[dest].Src)
+	}
+
+	for _, dst := range sets.StringKeySet(oldData.data).List() {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: this is the list of keys, which are in fact the destination, see the description of the `ConfigSyncData` at the beginning of the file

File: pkg/operator2/datasync/syncdata.go [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+		if oldData.data[dst].Resource == ConfigMapType {
+			oldConfigMapNames.Insert(dst)
+		} else {
+			oldSecretNames.Insert(dst)
+		}
+	}
+
+	notInUseConfigMapNames := oldConfigMapNames.Difference(newConfigMapNames)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: lol, yeah

File: pkg/operator2/datasync/syncdata.go [hunk 33]
```diff
@@ -0,0 +1,238 @@
...
+	}
+	for dest := range notInUseSecretNames {
+		syncOrDie(resourceSyncer.SyncSecret, dest, "")
+	}
+}
+
+// newSourceDataIDP returns a name which is unique amongst the IdPs, and sourceData
+// which describes the volumes and mount volumes to mount the CM/Secret to
+func newSourceDataIDP(index int, resourceType ResourceType, resourceName, field, key string) (string, sourceData) {
+	dest := getIDPName(index, field)
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go:82]: I'd still go with `dest`, it's the name of the destination resource

File: pkg/operator2/datasync/syncdata.go [hunk 34]
```diff
@@ -0,0 +1,238 @@
...
+	dirPath := getIDPPath(index, string(resourceType), dest)
+
+	ret := sourceData{
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: aye aye, returning `ret` directly, kept `dest` and `dirPath` variables for verbosity

File: pkg/operator2/datasync/syncdata.go (hunk 12 above)
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: addressed

File: pkg/operator2/datasync/syncdata.go (hunk 13 above)
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: done, exported the sync function and reused

File: pkg/operator2/datasync/syncdata.go (hunk 14 above)
Comment (stlaz) [reply pkg/operator2/datasync/syncdata.go]: You're right, I was blindly refactoring the old code without thinking about the names too much, but these name suggestion do make the code more readable, addressed :+1:

File: pkg/operator2/datasync/validation.go (hunk 17 above)
Comment (stlaz) [reply pkg/operator2/datasync/validation.go:39]: I don't mind the panic in case the programmer forgets to add a validator

File: pkg/operator2/datasync/validation.go (hunk 20 above)
Comment (stlaz) [reply pkg/operator2/datasync/validation.go]: added

File: pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go [hunk 35]
```diff
@@ -0,0 +1,82 @@
...
+	"k8s.io/client-go/tools/cache"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/controllers/configobservation"
+)
+
+func TestObserveTokenConfig(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go:18]: added 3 tests, 2 of them are more or less "the universe is sane", but they do test the corner cases

**Comment by stlaz** [comment]
addressed the comments, during which I noticed default authz token max age, which is not actually being dynamically set, was missing, fixed that

File: pkg/controllers/configobservation/oauth/observe_idps.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (marun) [reply pkg/controllers/configobservation/oauth/observe_idps.go:19]: Why should the pruning happen for anything but the existing config? Isn't newly created config always just the field(s) required?

File: pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go [hunk 36]
```diff
@@ -34,6 +34,52 @@ func TestObserveTokenConfig(t *testing.T) {
...
+					TokenConfig: configv1.TokenConfig{
+						AccessTokenMaxAgeSeconds: 0,
+					},
+				},
+			},
+			previouslyObservedConfig: map[string]interface{}{},
+			expected: map[string]interface{}{
+				"oauthConfig": map[string]interface{}{
+					"tokenConfig": map[string]interface{}{
+						"accessTokenMaxAgeSeconds":    float64(86400),
```
Comment (marun) [inline pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go:56]: (No action required) Maybe this is just my ignorance of accepted practice for cluster operators, but I find it strange to see magic numbers rather than constants.

File: pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go [hunk 37]
```diff
@@ -34,6 +34,52 @@ func TestObserveTokenConfig(t *testing.T) {
...
+						"authorizeTokenMaxAgeSeconds": float64(300),
+					},
+				},
+			},
+			errors: []error{},
+		},
+		{
+			name: "inactivity 0 means disabled",
```
Comment (marun) [inline pkg/controllers/configobservation/oauth/observe_tokenconfig_test.go:64]: I'm confused - how is this different from `"max age 0 still means default max age"`?

**Comment by marun** [comment]
@stlaz Thank you for cleanup! Other than the seemingly duplicated test, lgtm. Please squash for merge.

File: pkg/controllers/configobservation/oauth/observe_idps.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (stlaz) [reply pkg/controllers/configobservation/oauth/observe_idps.go:19]: that was in the case we got rid of the pruning, it needs to happen for any output and especially for the existing config, perhaps bad wording on my side

**Comment by stlaz** [comment]
/retest
unrelated failures
